core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1939,24 +1941,22 @@ private[spark] class DAGScheduler(
       hostToUnregisterOutputs: Option[String],
       maybeEpoch: Option[Long] = None): Unit = {
     val currentEpoch = maybeEpoch.getOrElse(mapOutputTracker.getEpoch)
+    logDebug(s"Removing executor $execId, fileLost: $fileLost, currentEpoch: $currentEpoch")
     if (!failedEpoch.contains(execId) || failedEpoch(execId) < currentEpoch) {
       failedEpoch(execId) = currentEpoch
       logInfo("Executor lost: %s (epoch %d)".format(execId, currentEpoch))
       blockManagerMaster.removeExecutor(execId)
-      if (fileLost) {
-        hostToUnregisterOutputs match {
-          case Some(host) =>
-            logInfo("Shuffle files lost for host: %s (epoch %d)".format(host, currentEpoch))
-            mapOutputTracker.removeOutputsOnHost(host)
-          case None =>
-            logInfo("Shuffle files lost for executor: %s (epoch %d)".format(execId, currentEpoch))
-            mapOutputTracker.removeOutputsOnExecutor(execId)
-        }
-        clearCacheLocs()
-
-      } else {
-        logDebug("Additional executor lost message for %s (epoch %d)".format(execId, currentEpoch))
+    }
+    if (fileLost && (!fileLostEpoch.contains(execId) || fileLostEpoch(execId) < currentEpoch)) {
```

Comment:
Can't we just use `if (fileLost)` here?

Reply:
No as we we would like to avoid calling `clearCacheLocs()` repeatedly for each fetch failure occurring within this stage. Fetch failures are coming from this map side executor but there could be several reducer side executors requesting block from that one and all would be generating a call here.

Reply:
As Attila explained, the purpose of fileLostEpoch is to track if this code has been called once for that executor and epoch, to avoid repeated calls to unregister the outputs, since there will usually be multiple fetch failures causing this method to be called.

Reply:
I see.